agent/system_metrics.go

```diff
@@ -0,0 +1,245 @@
+	system_metrics "github.com/evergreen-ci/timber/system_metrics"
+	"github.com/mongodb/ftdc"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+)
+
+type MetricCollector interface {
+	Name() string
+	Metadata() interface{}
+	Collect() (interface{}, error)
+}
+
+type SystemMetricsCollector struct {
+	mu          sync.Mutex
+	wg          sync.WaitGroup
+	innerCancel context.CancelFunc
+	logger      client.LoggerProducer
+	taskOpts    *system_metrics.SystemMetricsOptions
+	connOpts    system_metrics.ConnectionOptions
+	interval    time.Duration
+	collectors  []MetricCollector
+	id          string
+	client      *system_metrics.SystemMetricsClient
+	errored     bool
+	closed      bool
+}
+
+func NewSystemMetricsCollector(ctx context.Context, logger client.LoggerProducer, interval time.Duration,
+	c client.Communicator, t *task.Task, collectors []MetricCollector) (*SystemMetricsCollector, error) {
+
+	if interval < 0 {
+		return nil, errors.New("interval cannot be negative")
+	}
+	if interval == 0 {
+		interval = time.Minute
+	}
+
+	taskOpts, err := getTaskInfo(t)
+	if err != nil {
+		return nil, errors.Wrap(err, "incomplete task metadata")
+	}
+
+	if len(collectors) == 0 {
+		return nil, errors.New("must provide at least one MetricCollector")
+	}
+
+	bi, err := c.GetBuildloggerInfo(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting cedar dial options")
+	}
+	dialOpts := timber.DialCedarOptions{
+		BaseAddress: bi.BaseURL,
+		RPCPort:     bi.RPCPort,
+		Username:    bi.Username,
+		Password:    bi.Password,
+		APIKey:      bi.APIKey,
+		Retries:     10,
+	}
+	connOpts := system_metrics.ConnectionOptions{
+		DialOpts: dialOpts,
+	}
+
+	return &SystemMetricsCollector{
+		logger:     logger,
+		interval:   interval,
+		collectors: collectors,
+		taskOpts:   taskOpts,
+		connOpts:   connOpts,
+	}, nil
+}
+
+func (s *SystemMetricsCollector) Start(ctx context.Context) {
+	catcher := grip.NewBasicCatcher()
+	defer func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.errored = true
+		err := catcher.Resolve()
+		s.logger.System().Error(message.WrapError(err, message.Fields{
+			"message":  "error setting up system metrics collection",
+			"id":       s.id,
+			"connOpts": s.connOpts,
+			"taskOpts": *s.taskOpts,
+		}))
+	}()
+
+	var err error
+	s.client, err = system_metrics.NewSystemMetricsClient(ctx, s.connOpts)
+	if err != nil {
+		catcher.Add(err)
+		return
+	}
+
+	s.id, err = s.client.CreateSystemMetricRecord(ctx, *s.taskOpts)
+	if err != nil {
+		catcher.Add(err)
+		return
+	}
+
+	innerCtx := context.Background()
+	innerCtx, s.innerCancel = context.WithCancel(innerCtx)
+
+	for _, collector := range s.collectors {
+		s.wg.Add(1)
+		stream, err := s.client.StreamSystemMetrics(innerCtx, s.id, collector.Name(), system_metrics.StreamOpts{})
+		if err != nil {
+			catcher.Add(err)
+			return
+		}
+		ftdcCollector := ftdc.NewStreamingCollector(1, stream)
```

Comment:
It seems like using 1 for maxSamples will flush every single time it receives any data. Do we want to buffer it or are we just going to rely on the cedar client's internal buffering?